src/main/org/deidentifier/arx/metric/v2/DomainShareReliable.java

```diff
@@ -0,0 +1,185 @@
+ * the hierarchy.
+ * 
+ * @author Raffael Bild
+ */
+public class DomainShareReliable implements Serializable {
+
+    /** SVUID. */
+    private static final long                            serialVersionUID = -396317436976075163L;
+
+    /** The value representing a non-existent entry. */
+    private static final BigFraction                     NOT_AVAILABLE    = null;
+
+    /** One share per attribute. */
+    private final BigFraction[]                          shares;
+
+    /** If an attribute exists with different shares on different generalization levels, store the share in this map: <code>(((long)value) << 32) | (level & 0xffffffffL) -> share </code>. */
+    private transient LongObjectOpenHashMap<BigFraction> duplicates;
+
+    /**
+     * Creates a new set of domain shares derived from the given attribute.
+     *
+     * @param rawHierarchy
+     * @param encodedValues
+     * @param encodedHierarchy
+     */
+    public DomainShareReliable(String[][] rawHierarchy, 
+                                   String[] encodedValues, 
+                                   int[][] encodedHierarchy) {
+
+        this.duplicates = new LongObjectOpenHashMap<BigFraction>();
+        this.shares = new BigFraction[encodedValues.length];
+        Arrays.fill(shares, NOT_AVAILABLE);
+        @SuppressWarnings("unchecked")
+        ObjectIntOpenHashMap<String>[] maps = new ObjectIntOpenHashMap[rawHierarchy[0].length];
+        for (int level = 0; level < maps.length; level++) {
+            maps[level] = new ObjectIntOpenHashMap<String>();
+        }
+
+        // First, compute the share for each generalization strategy
+        for (int value = 0; value < rawHierarchy.length; value++) {
+            String[] transformation = rawHierarchy[value];
+            for (int level = 0; level < transformation.length; level++) {
+                ObjectIntOpenHashMap<String> map = maps[level];
+                String key = transformation[level];
+                if (!map.containsKey(key)) {
+                    map.put(key, 0);
+                }
+                map.put(key, map.get(key) + 1);
+            }
+        }
+
+        // Now transform into an array representation and handle duplicates
+        for (int row = 0; row < encodedHierarchy.length; row++) {
+            
+            int[] strategy = encodedHierarchy[row];
+            
+            for (int level = 0; level < strategy.length; level++){
+                
+                ObjectIntOpenHashMap<String> map = maps[level];
+                int value = strategy[level];
+                String keyString = encodedValues[value];
+                BigFraction share = new BigFraction(map.get(keyString), rawHierarchy.length);
+                BigFraction stored = shares[value];
+
+                // If duplicate
+                if (stored != NOT_AVAILABLE) {
+
+                    // If same share, simply continue
+                    if (stored == share) {

```

Comment:
Please double check. I am not sure(!), but stored.equals(share) may be what you are looking for. This compares references.